javaslang/src/main/java/javaslang/collection/AbstractMap.java

```diff
@@ -21,238 +21,238 @@
  *
  * @param <K> Key type
  * @param <V> Value type
- * @param <M> Map type
+ * @param <R> Map type
  * @author Ruslan Sennov
  * @since 2.0.0
  */
-abstract class AbstractMap<K, V, M extends AbstractMap<K, V, M>> implements Map<K, V> {
+abstract class AbstractMap<K, V, R extends AbstractMap<K, V, R>> implements Map<K, V> {
 
     private static final long serialVersionUID = 1L;
 
-    abstract M createFromEntries(Iterable<? extends Tuple2<? extends K, ? extends V>> entries);
+    abstract R createFromEntries(Iterable<? extends Tuple2<? extends K, ? extends V>> entries);
 
     /**
      * Returns an empty version of this traversable, i.e. {@code empty().isEmpty() == true}.
      *
      * @return an empty instance of this Map.
      */
-    abstract M emptyInstance();
+    abstract R emptyInstance();
 
     @SuppressWarnings("unchecked")
     @Override
-    public M put(Tuple2<? extends K, ? extends V> entry) {
+    public R put(Tuple2<? extends K, ? extends V> entry) {
         Objects.requireNonNull(entry, "entry is null");
-        return (M) put(entry._1, entry._2);
+        return (R) put(entry._1, entry._2);
     }
 
     @SuppressWarnings("unchecked")
     @Override
-    public M distinct() {
-        return (M) this;
+    public R distinct() {
+        return (R) this;
     }
 
     @Override
-    public M distinctBy(Comparator<? super Tuple2<K, V>> comparator) {
+    public R distinctBy(Comparator<? super Tuple2<K, V>> comparator) {
         Objects.requireNonNull(comparator, "comparator is null");
         return createFromEntries(iterator().distinctBy(comparator));
     }
 
     @Override
-    public <U> M distinctBy(Function<? super Tuple2<K, V>, ? extends U> keyExtractor) {
+    public <U> R distinctBy(Function<? super Tuple2<K, V>, ? extends U> keyExtractor) {
         Objects.requireNonNull(keyExtractor, "keyExtractor is null");
         return createFromEntries(iterator().distinctBy(keyExtractor));
     }
 
     @SuppressWarnings("unchecked")
     @Override
-    public M drop(long n) {
+    public R drop(long n) {
         if (n <= 0) {
-            return (M) this;
-        }
-        if (n >= length()) {
+            return (R) this;
+        } else if (n >= length()) {
             return emptyInstance();
+        } else {
```

Comment:
`else if` instead of `if` signals to the user that the branches are mutually exclusive